File: apps/easel/src/scheduler.rs
use std::sync::Arc;
use std::time::Duration;

use chrono::{Duration as ChronoDuration, NaiveDate, TimeZone, Utc};
use chrono_tz::Tz;

use crate::aic;
use crate::db::{self, DailyArtwork};
use crate::server::AppState;

pub async fn run(state: Arc<AppState>) {
    if let Err(e) = ensure_day(&state, &today_in_tz(&state.tz)).await {
        tracing::warn!("startup ensure_day failed: {e}");
    }

    if state.backfill_days > 0 {
        let dates = past_n_dates(&state.tz, state.backfill_days);
        match db::missing_dates(&state.db, &dates) {
            Ok(missing) => {
                tracing::info!(
                    "backfill: {} of last {} days missing",
                    missing.len(),
                    state.backfill_days
                );
                for d in missing {
                    if let Err(e) = ensure_day(&state, &d).await {
                        tracing::warn!("backfill {d} failed: {e}");
                    }
                }
            }
            Err(e) => tracing::error!("backfill missing_dates failed: {e}"),
        }
    }

    loop {
        let dur = duration_until_next_midnight(&state.tz);
        tracing::info!(
            "scheduler sleeping for {}s until next midnight in {}",
            dur.as_secs(),
            state.tz.name()
        );
        tokio::time::sleep(dur).await;
        let date = today_in_tz(&state.tz);
        if let Err(e) = ensure_day(&state, &date).await {
            tracing::warn!("scheduled ensure_day {date} failed: {e}");
        }
    }
}

pub async fn ensure_day(state: &AppState, date: &str) -> Result<(), String> {
    if db::get_daily(&state.db, date)
        .map_err(|e| e.to_string())?
        .is_some()
    {
        return Ok(());
    }
    let raw = aic::pick_unique(
        &state.http,
        &state.db,
        &state.classifications,
        state.max_dedup_retries,
    )
    .await?;
    let now = Utc::now().to_rfc3339();
    let daily: DailyArtwork = aic::raw_to_daily(raw, date.to_string(), now)
        .ok_or_else(|| "missing image_id on selected artwork".to_string())?;
    db::insert_daily(&state.db, &daily).map_err(|e| e.to_string())?;
    tracing::info!(
        "stored artwork {} for {} (image_id={})",
        daily.artwork_id,
        date,
        daily.image_id
    );
    Ok(())
}

pub fn today_in_tz(tz: &Tz) -> String {
    Utc::now()
        .with_timezone(tz)
        .date_naive()
        .format("%Y-%m-%d")
        .to_string()
}

pub fn past_n_dates(tz: &Tz, n: u32) -> Vec<String> {
    let today = Utc::now().with_timezone(tz).date_naive();
    (1..=n as i64)
        .filter_map(|i| today.checked_sub_signed(ChronoDuration::days(i)))
        .map(|d| d.format("%Y-%m-%d").to_string())
        .collect()
}

pub fn parse_date(s: &str) -> Option<NaiveDate> {
    NaiveDate::parse_from_str(s, "%Y-%m-%d").ok()
}

fn duration_until_next_midnight(tz: &Tz) -> Duration {
    let now = Utc::now().with_timezone(tz);
    let next_day = now.date_naive() + ChronoDuration::days(1);
    let next_midnight = tz
        .from_local_datetime(&next_day.and_hms_opt(0, 0, 1).expect("valid time"))
        .single()
        .or_else(|| {
            tz.from_local_datetime(&next_day.and_hms_opt(0, 0, 1).expect("valid time"))
                .earliest()
        })
        .unwrap_or_else(|| now + ChronoDuration::days(1));
    let delta = next_midnight.signed_duration_since(now);
    delta
        .to_std()
        .unwrap_or_else(|_| Duration::from_secs(60 * 60))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn past_n_dates_count() {
        let tz: Tz = "UTC".parse().unwrap();
        let dates = past_n_dates(&tz, 5);
        assert_eq!(dates.len(), 5);
    }

    #[test]
    fn past_n_dates_excludes_today() {
        let tz: Tz = "UTC".parse().unwrap();
        let today = today_in_tz(&tz);
        let dates = past_n_dates(&tz, 3);
        assert!(!dates.contains(&today));
    }

    #[test]
    fn parse_date_valid_invalid() {
        assert!(parse_date("2024-05-01").is_some());
        assert!(parse_date("2024-13-01").is_none());
        assert!(parse_date("notadate").is_none());
    }
}

File: apps/easel/scheduler.go
package main

import (
	"context"
	"time"
)

func (a *App) todayInTZ() string {
	return time.Now().In(a.TZ).Format("2006-01-02")
}

func (a *App) pastNDates(n int) []string {
	today := time.Now().In(a.TZ).Truncate(24 * time.Hour)
	out := make([]string, 0, n)
	for i := 1; i <= n; i++ {
		out = append(out, today.AddDate(0, 0, -i).Format("2006-01-02"))
	}
	return out
}

func parseDate(s string) (time.Time, bool) {
	t, err := time.Parse("2006-01-02", s)
	return t, err == nil
}

func (a *App) ensureDay(ctx context.Context, date string) error {
	if d, err := getDaily(a.DB, date); err != nil {
		return err
	} else if d != nil {
		return nil
	}
	raw, err := pickUnique(ctx, a.HTTP, a.DB, a.Classifications, a.ExcludeTerms, a.MaxDedupRetries)
	if err != nil {
		return err
	}
	now := time.Now().UTC().Format(time.RFC3339)
	daily := rawToDaily(raw, date, now)
	if daily == nil {
		return errMissingImage
	}
	if _, err := insertDaily(a.DB, daily); err != nil {
		return err
	}
	a.Log.Info("stored artwork", "artwork_id", daily.ArtworkID, "date", date, "image_id", daily.ImageID)
	return nil
}

var errMissingImage = &simpleError{"missing image_id on selected artwork"}

type simpleError struct{ msg string }

func (e *simpleError) Error() string { return e.msg }

func (a *App) runScheduler(ctx context.Context) {
	if err := a.ensureDay(ctx, a.todayInTZ()); err != nil {
		a.Log.Warn("startup ensure_day failed", "err", err)
	}
	if a.BackfillDays > 0 {
		dates := a.pastNDates(a.BackfillDays)
		missing, err := missingDates(a.DB, dates)
		if err != nil {
			a.Log.Error("backfill missing_dates failed", "err", err)
		} else {
			a.Log.Info("backfill", "missing", len(missing), "window_days", a.BackfillDays)
			for _, d := range missing {
				if err := a.ensureDay(ctx, d); err != nil {
					a.Log.Warn("backfill day failed", "date", d, "err", err)
				}
			}
		}
	}
	for {
		dur := a.durationUntilNextMidnight()
		a.Log.Info("scheduler sleeping", "seconds", dur.Seconds(), "tz", a.TZName)
		select {
		case <-ctx.Done():
			return
		case <-time.After(dur):
		}
		if err := a.ensureDay(ctx, a.todayInTZ()); err != nil {
			a.Log.Warn("scheduled ensure_day failed", "err", err)
		}
	}
}

func (a *App) durationUntilNextMidnight() time.Duration {
	now := time.Now().In(a.TZ)
	nextDay := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 1, 0, a.TZ).AddDate(0, 0, 1)
	delta := nextDay.Sub(now)
	if delta < time.Second {
		return time.Hour
	}
	return delta
}
